web/src/main/scala/quasar/api/services/data.scala

```diff
@@ -0,0 +1,114 @@
+import org.http4s.server._
+import pathy.Path._
+import quasar.repl.Prettify
+import quasar.Predef._
+import quasar.api.{MessageFormat, AsPathyPath}
+import quasar.fp._
+import quasar.fs._
+import scodec.bits.ByteVector
+
+import scalaz._
+import scalaz.concurrent.Task
+import scalaz.stream.Process
+import scalaz.syntax.validation._
+import scalaz.syntax.traverse._
+import scalaz.syntax.TraverseOps
+import scalaz.syntax.monad._
+import scalaz.syntax.show._
+import scalaz.syntax.std.option._
+import scalaz.std.option._
+
+object data {
+  import FileSystemError._
+  import scalaz.Validation.FlatMap._
+
+  implicit val naturalParamDecoder: org.http4s.QueryParamDecoder[Natural] = new QueryParamDecoder[Natural] {
+    def decode(value: QueryParameterValue): ValidationNel[ParseFailure, Natural] =
+      QueryParamDecoder[Long].decode(value).flatMap(long =>
+        Natural(long).toSuccess(NonEmptyList(ParseFailure(value.value, "must be >= 0")))
+      )
+  }
+
+  implicit val positiveParamDecoder: org.http4s.QueryParamDecoder[Positive] = new QueryParamDecoder[Positive] {
+    def decode(value: QueryParameterValue): ValidationNel[ParseFailure, Positive] =
+      QueryParamDecoder[Long].decode(value).flatMap(long =>
+        Positive(long).toSuccess(NonEmptyList(ParseFailure(value.value, "must be >= 1")))
+      )
+  }
+
+  object Offset extends OptionalValidatingQueryParamDecoderMatcher[Natural]("offset")
+  object Limit  extends OptionalValidatingQueryParamDecoderMatcher[Positive]("limit")
+
+  type FSTask[A] = FileSystemErrT[Task, A]
+
+  /** Flatten by inserting the [[quasar.fs.FileSystemError]] into the failure case of the [[scalaz.concurrent.Task]] */
+  val flatten = new (FSTask ~> Task) {
+    def apply[A](t: FSTask[A]): Task[A] =
+      t.fold(e => Task.fail(new RuntimeException(e.shows)), Task.now).join
+  }
+
+  def service[S[_]: Functor](f: S ~> Task)(implicit R: ReadFile.Ops[S],
+                                                    W: WriteFile.Ops[S],
+                                                    M: ManageFile.Ops[S]): HttpService = {
+
+    HttpService {
+      case req @ GET -> AsPathyPath(path) :? Offset(offsetParam) +& Limit(limitParam) => {
+        val offsetWithDefault = offsetParam.getOrElse(Natural._0.successNel)
+        val offsetWithErrorMsg: String \/ Natural = offsetWithDefault.disjunction.leftMap(
+          nel => s"invalid offset: ${nel.head.sanitized} (${nel.head.details})")
+        val limitWithErrorMsg: String \/ Option[Positive] = limitParam.traverseU(_.disjunction.leftMap(
+          nel => s"invalid limit: ${nel.head.sanitized} (${nel.head.details})"))
+        val possibleResponse = for {
+          offset <- offsetWithErrorMsg
+          limit <- limitWithErrorMsg
+        } yield {
+          val requestedFormat = MessageFormat.fromAccept(req.headers.get(Accept))
+          def formatAsHttpResponse[A: EntityEncoder](data: Process[R.M, A], contentType: `Content-Type`): Task[Response] = {
```

Comment:
Can we pull this method out, either above HttpService or all the way out as a sibling to `zippedBytes`? It clutters up the flow of the handler.